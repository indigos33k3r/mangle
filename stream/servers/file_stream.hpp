#ifndef MANGLE_STREAM_FILESERVER_H
#define MANGLE_STREAM_FILESERVER_H

#include "std_stream.hpp"
#include <fstream>
#include <stdexcept>

namespace Mangle {
namespace Stream {

/** Very simple file input stream, based on std::ifstream
 */
class FileStream : public StdStream
{
  std::ifstream file;

 public:
  FileStream(const std::string &name)
    : StdStream(&file)
    {
      file.open(name.c_str(), std::ios::binary);

      if(file.fail())
        throw std::runtime_error("FileStream: failed to open file " + name);
    }
  ~FileStream() { file.close(); }

  static StreamPtr Open(const std::string &name)
  { return StreamPtr(new FileStream(name)); }

  // Convenience function to read a buffer directly from a file, then
  // close it
  static size_t Read(const std::string &file, void *ptr, size_t count)
  { return Open(file)->read(ptr, count); }
};

typedef boost::shared_ptr<FileStream> FileStreamPtr;

}} // namespaces
#endif
