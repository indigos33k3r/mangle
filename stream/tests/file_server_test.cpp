#include "../servers/file_stream.hpp"
#include <iostream>

using namespace Mangle::Stream;
using namespace std;

int main()
{
  StreamPtr inp = FileStream::Open("file_server_test.cpp");

  char buf[21];
  buf[20] = 0;
  cout << "pos=" << inp->tell() << " eof=" << inp->eof() << endl;
  inp->read(buf, 20);
  cout << "First 20 bytes: " << buf << endl;
  cout << "pos=" << inp->tell() << " eof=" << inp->eof() << endl;
  return 0;
}
