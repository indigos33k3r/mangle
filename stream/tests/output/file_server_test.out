pos=0 eof=0
First 20 bytes: #include "../servers
pos=20 eof=0
First char of file: #
